### PR TITLE
servicecontroller fixes: exponential backoff

### PR DESCRIPTION
**Pull request by justinsb**
Edit: Works in my testing now

Issue #21952

### CODE REVIEW

**Comment by justinsb** [comment]
cc @thockin @bprashanth 

Fixes that I think we could safely apply to 1.2; that I think/hope will solve the worst of the problems, and give us breathing room to re-address in 1.3

**Comment by a-robinson** [comment]
I like the backoff changes, and I think that the resourceVersion checking will prevent permanent bad states in most cases, but I don't think that this will handle all cases involving deletes.

For a delete, we remove the cachedService from the cache rather than updating its appliedState (because we don't have a resourceVersion to update it to AFAIK). I think this could lead to problems of reordering where a cluster update could fail, then we'd delete the load balancer and service, then the update would be retried, recreating the load balancer. If the retry succeeds, a sync of all known services would presumably resolve this, but we could end up retrying the update forever if the retries never succeed.

A simple, standard way to fix this would be to just re-GET the object at the start of each processing attempt rather than relying on the object in the delta. Would you mind swapping over to that (and I'll abandon #19879), or should I update #19879?

**Comment by justinsb** [comment]
@a-robinson so many edge cases :-)  I agree that a GET would be clean, though I'm not exactly sure how to go about it.  Would you mind doing that in #19879 or a new PR)?  Then I can extract just the backoff piece of this PR after yours has merged.  I think this sequence should be easiest on everyone!

**Comment by justinsb** [comment]
@a-robinson just the backoff, now that #22069 has merged

**Comment by justinsb** [comment]
@k8s-bot test this github flake #21995

File: pkg/controller/service/servicecontroller.go [hunk 1]
```diff
@@ -244,11 +255,11 @@ func (s *ServiceController) processDelta(delta *cache.Delta) (error, bool) {
 		if err != nil {
 			message := "Error deleting load balancer (will retry): " + err.Error()
 			s.eventRecorder.Event(deltaService, api.EventTypeWarning, "DeletingLoadBalancerFailed", message)
-			return err, retryable
+			return err, cachedService.nextRetryDelay()
 		}
 		s.eventRecorder.Event(deltaService, api.EventTypeNormal, "DeletedLoadBalancer", "Deleted load balancer")
 		s.cache.delete(namespacedName.String())
-		return nil, notRetryable
+		return nil, cachedService.nextRetryDelay()
```
Comment (a-robinson) [inline pkg/controller/service/servicecontroller.go]: Shouldn't you reset the retry delay and return doNotRetry here? This is a success case

**Comment by a-robinson** [comment]
Looks great other than the one comment

**Comment by justinsb** [comment]
@a-robinson great spot - you are 100% right.  Pushed a fixed version.

**Comment by a-robinson** [comment]
LGTM

**Comment by eparis** [comment]
@k8s-bot test this, issue #22435

**Comment by justinsb** [comment]
Marking this P1 ... this (along with @a-robinson's fix to fix the underlying problem) will resolve a bunch of AWS issues targeted for 1.2

**Pull request merged**